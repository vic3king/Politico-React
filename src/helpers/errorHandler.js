const errorHandler = obj => {
  let error;

  if (typeof obj === 'string') {
    error = obj;
  }

  if (typeof obj === 'object') {
    Object.values(obj).forEach(err => {
      error = err;
    });
  }

  return error;
};

export default errorHandler;
