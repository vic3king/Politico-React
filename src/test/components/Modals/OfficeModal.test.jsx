import React from 'react';
import { shallow } from 'enzyme';
import { notify } from 'react-notify-toast';
import OffficeModal from '../../../components/Modals/OfficeModal';
import Office from '../../../services/offices';
import errorHandler from '../../../helpers/errorHandler';

jest.mock('../../../services/offices');
jest.mock('react-notify-toast');

const hide = jest.fn();
const updateOfficesState = jest.fn();

let wrapper;
describe('InterestFormModal component', () => {
  global.fetch = jest.fn();
  global.localStorage.setItem('user', '{}');

  beforeEach(() => {
    wrapper = shallow(
      <OffficeModal hide={hide} updateOfficesState={updateOfficesState} />
    );
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('hide method', () => {
    wrapper = shallow(
      <OffficeModal hide={hide} updateOfficesState={updateOfficesState} />
    );
    const instance = wrapper.instance();
    let event;
    it('should call preventDefault on event', async () => {
      event = {
        preventDefault: jest.fn(),
      };

      await instance.hide(event);
      expect(event.preventDefault).toBeCalled();
    });
  });

  describe('onInputChange method', () => {
    it('should update formData state ', () => {
      const instance = wrapper.instance();
      const event = {
        preventDefault: jest.fn(),
        target: { id: 'office' },
      };
      const formDetails = {
        name: 'test',
      };
      instance.setState({
        formDetails,
      });

      instance.onInputChange(event);

      expect(instance.state.formDetails).toEqual({
        name: 'test',
      });
    });
  });

  describe('onButtonSubmit method', () => {
    let instance;
    let event;

    describe('Api call success', () => {
      beforeAll(() => {
        Office.postOffice = jest.fn().mockImplementation(() =>
          Promise.resolve({
            status: 201,
          })
        );
      });
      beforeEach(() => {
        instance = wrapper.instance();
        event = {
          preventDefault: jest.fn(),
        };
      });

      it('should call preventDefault on event', async () => {
        await instance.onButtonSubmit(event);

        expect(event.preventDefault).toBeCalled();
      });
      it('should call toast the right message on api call success', async () => {
        await instance.onButtonSubmit(event);
        expect(notify.show).toBeCalledWith('success');
      });
      it('should set loading state to false after successfull api call is made', async () => {
        await instance.onButtonSubmit(event);
        expect(instance.state.loading).toBeFalsy();
      });
      it('should call the updateOfficesState method with id from api call', async () => {
        await instance.onButtonSubmit(event);
        expect(updateOfficesState).toBeCalled();
      });
    });

    describe('Api call 400 errors', () => {
      beforeAll(() => {
        Office.postOffice = jest.fn().mockImplementation(() =>
          Promise.resolve({
            status: 400,
            error: [{ error: 'validation error' }],
          })
        );
      });
      beforeEach(() => {
        instance = wrapper.instance();
        event = {
          preventDefault: jest.fn(),
        };
      });

      it('should set loading state to false on validation error', async () => {
        await instance.onButtonSubmit(event);
        expect(instance.state.loading).toBeFalsy();
      });

      it('should toast the error message', async () => {
        await instance.onButtonSubmit(event);
        expect(notify.show).toBeCalledWith(
          errorHandler('validation error'),
          'error'
        );
      });
    });
  });
});
