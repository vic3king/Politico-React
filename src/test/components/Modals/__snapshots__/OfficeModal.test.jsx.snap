// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InterestFormModal component should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <OffficeModal
    hide={[MockFunction]}
    updateOfficesState={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <$class
          options={Object {}}
        />,
        false,
        <Modal
          ButtonOne={
            <Button
              className="cte-off"
              id="office-btn"
              onClick={[Function]}
              type="submit"
              value="Create office"
            />
          }
          CloseModal={
            <button
              id="btn"
              onClick={[Function]}
              type="button"
            >
              X
            </button>
          }
          Header={
            <Header
              text="Create a Political Office:"
            />
          }
          InputOne={
            <Input
              id="name"
              onChange={[Function]}
              placeholder="e.g Local government Chairman"
              required={true}
              type="text"
              value=""
            />
          }
          InputTwo={
            <Input
              id="ageLimit"
              onChange={[Function]}
              placeholder=""
              required={true}
              type="number"
              value=""
            />
          }
          Select={
            <Select
              onChange={[Function]}
            />
          }
          SpanOne={
            <Span
              text="Office Type: "
            />
          }
          SpanThree={
            <Span
              text="AgeLimit"
            />
          }
          SpanTwo={
            <Span
              text="Office: "
            />
          }
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "options": Object {},
        },
        "ref": null,
        "rendered": null,
        "type": [MockFunction],
      },
      false,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "ButtonOne": <Button
            className="cte-off"
            id="office-btn"
            onClick={[Function]}
            type="submit"
            value="Create office"
          />,
          "CloseModal": <button
            id="btn"
            onClick={[Function]}
            type="button"
          >
            X
          </button>,
          "Header": <Header
            text="Create a Political Office:"
          />,
          "InputOne": <Input
            id="name"
            onChange={[Function]}
            placeholder="e.g Local government Chairman"
            required={true}
            type="text"
            value=""
          />,
          "InputTwo": <Input
            id="ageLimit"
            onChange={[Function]}
            placeholder=""
            required={true}
            type="number"
            value=""
          />,
          "Select": <Select
            onChange={[Function]}
          />,
          "SpanOne": <Span
            text="Office Type: "
          />,
          "SpanThree": <Span
            text="AgeLimit"
          />,
          "SpanTwo": <Span
            text="Office: "
          />,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <$class
            options={Object {}}
          />,
          false,
          <Modal
            ButtonOne={
              <Button
                className="cte-off"
                id="office-btn"
                onClick={[Function]}
                type="submit"
                value="Create office"
              />
            }
            CloseModal={
              <button
                id="btn"
                onClick={[Function]}
                type="button"
              >
                X
              </button>
            }
            Header={
              <Header
                text="Create a Political Office:"
              />
            }
            InputOne={
              <Input
                id="name"
                onChange={[Function]}
                placeholder="e.g Local government Chairman"
                required={true}
                type="text"
                value=""
              />
            }
            InputTwo={
              <Input
                id="ageLimit"
                onChange={[Function]}
                placeholder=""
                required={true}
                type="number"
                value=""
              />
            }
            Select={
              <Select
                onChange={[Function]}
              />
            }
            SpanOne={
              <Span
                text="Office Type: "
              />
            }
            SpanThree={
              <Span
                text="AgeLimit"
              />
            }
            SpanTwo={
              <Span
                text="Office: "
              />
            }
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "options": Object {},
          },
          "ref": null,
          "rendered": null,
          "type": [MockFunction],
        },
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "ButtonOne": <Button
              className="cte-off"
              id="office-btn"
              onClick={[Function]}
              type="submit"
              value="Create office"
            />,
            "CloseModal": <button
              id="btn"
              onClick={[Function]}
              type="button"
            >
              X
            </button>,
            "Header": <Header
              text="Create a Political Office:"
            />,
            "InputOne": <Input
              id="name"
              onChange={[Function]}
              placeholder="e.g Local government Chairman"
              required={true}
              type="text"
              value=""
            />,
            "InputTwo": <Input
              id="ageLimit"
              onChange={[Function]}
              placeholder=""
              required={true}
              type="number"
              value=""
            />,
            "Select": <Select
              onChange={[Function]}
            />,
            "SpanOne": <Span
              text="Office Type: "
            />,
            "SpanThree": <Span
              text="AgeLimit"
            />,
            "SpanTwo": <Span
              text="Office: "
            />,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
